### PR TITLE
[auth] Request JWT secret from Kong

### PR DESCRIPTION
**Pull request by jaylees14**
Rather than managing secret keys to sign JWTs ourselves, we can leverage additional Kong functionality. This has several advantages:
- Kong will automatically require and validate JWTs before calling any other service
- Kong controls the secure stuff, so we don't have to
- Less code ❤️ 

So... to do this, the Auth service has to do several things (followed from https://docs.konghq.com/hub/kong-inc/jwt/)
- Register as a "consumer"
- Use this registration to request a secret & key to sign JWTs with

To do this, the auth service requires access to Kong's "Admin" port, exposed on 8001. This would normally _not_ be exposed to the general public, however we currently have it configured that way for ease of testing / configuration.

One issue we now have is that the auth service depends on the Kong service already being ready and active. This **isn't** possible from the Docker Compose `depends-on`(https://docs.docker.com/compose/compose-file/#depends_on), so we are forced to use a shell script that periodically checks for readiness, following the advice from https://docs.docker.com/compose/startup-order/. 

### Test Plan
First spin up all the containers and configure Kong:
```
 ❯❯❯ docker-compose up --build  
 ❯❯❯ sh kong/configure-kong.sh
```

Now try and make a request to a service that isn't auth:
```
❯❯❯ curl -X GET localhost:8000/api/user/2
{"message":"Unauthorized"}
```

Now create a new user:
```
❯❯❯ curl -X POST localhost:8000/api/auth -d '{"email":"jay@test.com", "password":"abcdefgh"}'
{"AccessToken":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImpheUB0ZXN0LmNvbSIsImV4cCI6MTU4Mjg5MzU3NywiaXNzIjoiMG9kVHc1d1FXOHRVdlUxVEM5elQxa2pjNnpKQVM1RVYifQ.s89vHXXsZ2xpiCdMGtpABcRA6kbCIlgCWxV3v9SekAI"}
```

Now make the same request as before, including the token in the headers - it works 😍 
```
❯❯❯ curl -X GET localhost:8000/api/user/2 -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImpheUB0ZXN0LmNvbSIsImV4cCI6MTU4Mjg5MzU3NywiaXNzIjoiMG9kVHc1d1FXOHRVdlUxVEM5elQxa2pjNnpKQVM1RVYifQ.s89vHXXsZ2xpiCdMGtpABcRA6kbCIlgCWxV3v9SekAI"
{"error":"user not found with ID 2"}
```

Now to make sure this is actually checking the JWT correctly, I've created two JWTs using the debugger at [jwt.io](jwt.io) - one that contains the correct `iss` tag for the `auth-service`, but incorrect signature, and one with random contents:

#### Correct `iss`, incorrect signature
```
❯❯❯ curl -X GET localhost:8000/api/user/2 -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJpc3MiOiIwb2RUdzV3UVc4dFV2VTFUQzl6VDFramM2ekpBUzVFViJ9.1bliwzt-ZllhodP0anzB94htvuFsDcykhzeuV_jRlNE"
{"message":"Invalid signature"}
```
Screenshot 2020-02-27 at 12 45 12

#### Random Contents:
```
 ❯❯❯ curl -X GET localhost:8000/api/user/2 -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJpc3MiOiJSQU5ET01QRVJTT04ifQ._Nv1aKd18JqKf6JVIeUd2-IWNFlPF2pURGRdN19UI1Y"
{"message":"No credentials found for given 'iss'"}
```
Screenshot 2020-02-27 at 12 43 24

### CODE REVIEW

File: auth/comm/handler.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+func requestCredential(hostname string, consumer *consumerResponse) (*JWTCredential, error) {
+	reqUrl := fmt.Sprintf("%s/consumers/%s/jwt", hostname, consumer.Username)
+	res, err := http.Post(reqUrl, "application/x-www-form-urlencoded", nil)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != 201 {
```
Comment (willjvsmith) [inline auth/comm/handler.go]: Perhaps replace `201` with the http enum

File: auth/comm/handler.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	postData := url.Values{}
+	postData.Set("username", "auth-service")
+
+	res, err := http.PostForm(fmt.Sprintf("%s/consumers", hostname), postData)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != 201 {
```
Comment (willjvsmith) [inline auth/comm/handler.go]: Same here

**Review by willjvsmith — APPROVED** [review]
LGTM 🚀

**Review by Lewiky — APPROVED** [review]
YAS QUEEN LGTM BOI Espanol Espanol Numbah One

**Pull request merged**